[#123362] Travel Pay, Complex claims initial redux setup

seesleestak opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to TeamSites and delete the rest of this section)
- [ ] Yes, I'm removing, renaming or moving a folder

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No
- [ ] Yes, and I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#that-sounds-normal-so-whats-the-proxy-all-about) to test the injected header scenario

## Summary

Adding redux integration for our initial mileage flow in complex claims. Also added some mock data for API calls leveraged by the new redux actions.

All of this work is behind the flipper `travel_pay_enable_complex_claims`

The entry point for local testing is something like: http://localhost:3001/my-health/travel-pay/file-new-claim/12345/45678/choose-expense

Also, I'm using the mock-api for local testing: `yarn mock-api --responses src/applications/travel-pay/services/mocks/index.js
`

## Related issue(s)

https://github.com/department-of-veterans-affairs/va.gov-team/issues/123362

## Testing done

* Path from `/choose-expense` to `/review`
* Deleting expense on `/review`
* Returning to edit mileage expense page from review

## Screenshots

_Note: The UX is not visually different, since most of this work was wiring up data._

https://github.com/user-attachments/assets/e6f0f691-1375-4b8c-b71e-db818f638664

## What areas of the site does it impact?

Travel Pay

## Acceptance criteria

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/prepare-for-an-accessibility-staging-review) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

seesleestak committed:
Add conditional route definitions

seesleestak committed:
Update URL references for complex claims

seesleestak committed:
Test updates

seesleestak committed:
Fix addMoreExpenses route

seesleestak committed:
More test refinement

seesleestak committed:
More test tweaks

seesleestak committed:
Add optional expense id for expense entry routes

seesleestak committed:
Add actions

seesleestak committed:
Add reducer state updates

seesleestak committed:
Add selectors

seesleestak committed:
Update wrapper with new action usage

seesleestak committed:
Create TravelPayButtonPair

seesleestak committed:
Initial Mileage implementation with redux

seesleestak committed:
Remove redirect logic

seesleestak committed:
Add some API mocks

seesleestak committed:
Initial implementation with redux for Review

seesleestak committed:
Test fixes

seesleestak committed:
Add action tests

pmclaren19 commented on src/applications/travel-pay/components/shared/TravelPayButtonPair.jsx:
```diff
@@ -0,0 +1,48 @@
+import PropTypes from 'prop-types';
+import { VaButton } from '@department-of-veterans-affairs/component-library/dist/react-bindings';
+
+const TravelPayButtonPair = ({
```
why is this component needed? can we jsut use va-button-pair?

seesleestak replied:
It's needed because button pair doesn't support a loading state.

seesleestak committed:
Resolve merge conflicts

seesleestak committed:
Mileage button pair refinement

seesleestak committed:
Add loading handling for expense deletion

seesleestak committed:
Fix cancel expense routing on mileage

seesleestak committed:
Casing tweak on delete expense

seesleestak committed:
Route fix in intro page

seesleestak committed:
Rework the store structure

seesleestak committed:
Fix dynamic id return in mocks

seesleestak committed:
Update Intro page with store usage

seesleestak committed:
Handle edit and loading states on Mileage

seesleestak committed:
Add more loading for wrapper calls

seesleestak committed:
Store usage and loading tweaks for Review

seesleestak committed:
Update tests

seesleestak committed:
Separate conditions for appointment and claims calls

seesleestak committed:
Test fixes after store restructure

seesleestak committed:
Test fixes

seesleestak committed:
Resolve merge conflicts

seesleestak committed:
Replace sign agreement move in test, update routing

seesleestak committed:
Resolve merge conflicts

seesleestak committed:
Use trip type and expense type constants

seesleestak committed:
Remove unused appointment info on mileage

seesleestak committed:
Tweaks to expense deletion loading

seesleestak committed:
Add fetch for claim details after expense operations

seesleestak committed:
Update usage

seesleestak committed:
Test fixes

seesleestak committed:
Wire up confirmation page

seesleestak committed:
Fix mock for comple claim submission

seesleestak committed:
Update store for complex claim submission

seesleestak committed:
Wire up implementation on agreement page for submission

seesleestak committed:
Don't use mock data on Review

seesleestak committed:
Ensure claim details are getting merged properly

seesleestak committed:
Margin tweak for confirmation alert

seesleestak committed:
Agreement page test updates

seesleestak committed:
Cypress test fix

seesleestak committed:
Hard code round trip

seesleestak committed:
Resolve merge conflicts

seesleestak commented on src/applications/travel-pay/components/complex-claims/pages/ReviewPage.jsx:
```diff
@@ -175,7 +174,11 @@ const ReviewPage = ({ claim, message }) => {
         <p>
           <strong>Total:</strong> $
-          {formatAmount(overriddenClaim.totalCostRequested)}
+          {formatAmount(
```
@pmclaren19 I only noticed this while working on this branch, but I think there may be some overlap between the `formatAmount` and the existing `currency` helper we have. I don't mind which we use but I imagine we only need one.

pmclaren19 replied:
will currency also make a null or undefined amount show 0.00 ? I just tried to go through the differences and thats the only think that i think looks different

seesleestak replied:
I would guess not, but do we _want_ missing values to display as 0? That seems like something that should be protected against before rendering.

seesleestak committed:
e2e test update

seesleestak committed:
Make sure claim details exist in complexClaim portion of the store

seesleestak committed:
Additional expense merging logic

seesleestak committed:
Override some mocks for mileage with complex claims

pmclaren19 commented on src/applications/travel-pay/redux/actions.js:
```diff
@@ -115,6 +137,17 @@ const submitClaimFailure = error => ({
   error,
 });
 
+// Submitting a complex travel claim
```
this should be moved down to where the other submit complex claim logic is

pmclaren19 commented on src/applications/travel-pay/redux/actions.js:
```diff
@@ -135,3 +168,246 @@ export function submitMileageOnlyClaim(appointmentData) {
+      };
+
+      const expenseUrl = `${environment.API_URL}/travel_pay/v0/expenses/${
+        EXPENSE_TYPES[expenseType]?.route
```
if we are going to do this then we will need to update the  EXPENSE_TYPES.route to match whats in the swagger doc. AirTravel, CommonCarrier are incorrect

seesleestak committed:
Relocate complex claim submit actions in file

seesleestak committed:
Add apiRoute to EXPENSE_TYPES

seesleestak committed:
Update api routes in actions

seesleestak committed:
Fix usage of apiRoute in consumers

kjduensing commented on src/applications/travel-pay/components/complex-claims/pages/AgreementPage.jsx:
```diff
@@ -1,24 +1,40 @@
       setIsAgreementError(false);
-      // TODO Add logic for Submitting the claim
-      navigate(`/file-new-claim/${apptId}/${claimId}/confirmation`);
+      try {
```
Nitpicky: The negatives here twist my brain a little. Could we do something like (either now or in the future):

```javascript
setIsAgreementError(!isAgreementChecked);

if (isAgreementChecked) {
  navigate(...);
  try {
    // ...
  catch (error) {
    // ...
  }
}
```

seesleestak committed:
Simplify agreement check logic on agreement page

seesleestak committed:
Add expense id and type check for expense actions

seesleestak committed:
Add tests

kjduensing commented on src/applications/travel-pay/redux/actions.js:
```diff
@@ -135,3 +156,272 @@ export function submitMileageOnlyClaim(appointmentData) {
+      try {
+        await dispatch(getComplexClaimDetails(claimId));
+      } catch (fetchError) {
+        // Silently continue if fetching details fails
```
Why silently continue?

seesleestak replied:
I understand "silently" may be triggering 😄

seesleestak committed:
Remove extra claimDetails?.totalCostRequested

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/AgreementPage.jsx:
```diff
@@ -1,24 +1,39 @@
+    if (isAgreementChecked) {
+      try {
+        // Submit the complex claim
+        await dispatch(submitComplexClaim(claimId, claimData));
```
🎊

pmclaren19 commented on src/applications/travel-pay/tests/components/complex-claims/pages/ConfirmationPage.unit.spec.jsx:
```diff
@@ -1,25 +1,63 @@
+import reducer from '../../../../redux/reducer';
 import ConfirmationPage from '../../../../components/complex-claims/pages/ConfirmationPage';
 
+const appointment = {
```
shouldnt this mock have the apptId in it?

seesleestak committed:
Put title check assertion back in confirmation page test

seesleestak committed:
Resolve merge conflicts

seesleestak committed:
Consolidate around TRIP_TYPES

seesleestak commented on src/applications/travel-pay/tests/components/complex-claims/pages/ExpenseCard.unit.spec.jsx:
```diff
@@ -157,16 +164,21 @@ describe('ExpenseCard', () => {
               />
             }
           />
-          <Route path="/mileage" element={<Mileage />} />
```
I removed the `editToRoute` since now the component itself can determine what that is instead of needing it passed down.

seesleestak replied:
Oh just because the test doesn't need to render the entire other page to know that the routing behavior is working correctly. Rendering anything larger than we need to just makes the tests slower.

pmclaren19 replied:
then we might want to consider updating other tests that are rendering these! - non blocking thought for later

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/Mileage.jsx:
```diff
@@ -114,51 +174,56 @@ const Mileage = () => {
         id="trip-type"
         onVaValueChange={handleTripTypeChange}
         value={tripType}
-        label="Which address did you depart from?"
```
good catch

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/Mileage.jsx:
```diff
@@ -114,51 +174,56 @@ const Mileage = () => {
+        </>
+      )}
+      <TravelPayButtonPair
+        continueText={expenseId ? 'Save and continue' : 'Continue'}
```
Looks like this is coming from these designs - https://www.figma.com/design/EM3ejHDOgnxaeOVRw6xnpf/Complex-travel-reimbursement-claims?node-id=8203-36649&t=NMxRVMJKcAgnagqN-0 👍

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/ReviewPage.jsx:
```diff
@@ -67,11 +75,9 @@ const ReviewPage = ({ claim, message }) => {
   const onClose = () => setVisible(false);
   const addMoreExpenses = () => {
     navigate(`/file-new-claim/${apptId}/${claimId}/choose-expense`);
-    // TODO Add logic to add more expenses
```
why did you remove this it doesnt look like the add expense logic was added yet

seesleestak replied:
We just route from to choose-expense from the review page. We only add the expense from one of those expense pages.

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/ExpenseCard.jsx:
```diff
@@ -1,102 +1,121 @@
-    // TODO Add logic for this function, it will delete the expense once the delete modal is confirmed
-    // eslint-disable-next-line no-console
-    console.log(`Delete clicked for expense id: ${expenseId}`);
+  const header = `${formatDate(expense.dateIncurred)}, ${currency(
```
👍

pmclaren19 commented on src/applications/travel-pay/sass/travel-pay.scss:
```diff
@@ -146,3 +146,55 @@ h2#pagination-info {
 }
+
+// Custom button pair implementation for travel pay - exact copy of usa-button-group
+.travel-pay-button-group {
```
can we improve this so you dont have to declare `.travel-pay-button-group` 2x? would this be the same thing:

```
.travel-pay-button-group {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  padding-left: 0;
  list-style-type: none;
  margin: 0 -0.25rem; // Handles item spacing uniformly

  @media (min-width: 30em) {
    flex-direction: row;
    flex-wrap: nowrap;
    align-items: stretch;
  }
}

seesleestak commented on src/applications/travel-pay/components/complex-claims/pages/Mileage.jsx:
```diff
@@ -114,51 +174,56 @@ const Mileage = () => {
-        onSecondaryClick={handleBack}
+      {!expenseId && (
+        <>
+          <VaModal
```
Sorry, I'm not following. Are you referring to the modal? Or to field validation?

pmclaren19 replied:
Nevermind after looking at the designs again I found the edit logic - https://www.figma.com/design/EM3ejHDOgnxaeOVRw6xnpf/Complex-travel-reimbursement-claims?node-id=8333-34508&t=RH5t9V29R4TtQWQO-0 .

So in summary if there is an expense id then the cancel modal will never be shown, especially since the edit designs dont have the `Cancel adding this mileage expense` button. Otherwise if there isnt a mileage expense than a user will see that button and when they click it they will see the cancel modal.

seesleestak replied:
Oh gotcha. Yeah, just taking away the cancel modal for the "edit" mode on the expense pages.

pmclaren19 reviewed (commented):
@seesleestak  I pulled down the branch and ran locally and found the following issues:
- The back button on the choose expense page isnt navigating me anywhere when i click it
- the mileage page is missing a << on the back button
- I attempted to add a parking expense but when I got to the review page I did not see the parking expense that I added I saw a total on 0. When I reloaded the page I saw a mileage expense of $1.16 . I attempted to add several other expenses and they did not work
- It looked like I was deleting the mileage expense but then it was re-added after I reloaded the page ... is this how it should act with mocks?
- Edit mileage appears to work, although I cant actually change any of the data

seesleestak committed:
Button pair content tweak

seesleestak commented:
> * The back button on the choose expense page isnt navigating me anywhere when i click it

This is interesting. I think it's because it _is_ working, routing you back to the intro page, and the intro page `useEffect` sees a claim id and routes you back to choose-expense. I'll think on it

> 
> * the mileage page is missing a << on the back button
> 

Updated 👍 

> * I attempted to add a parking expense but when I got to the review page I did not see the parking expense that I added I saw a total on 0. When I reloaded the page I saw a mileage expense of $1.16 . I attempted to add several other expenses and they did not work
> 
> * It looked like I was deleting the mileage expense but then it was re-added after I reloaded the page ... is this how it should act with mocks?
> 
> * Edit mileage appears to work, although I cant actually change any of the data

Yeah, all of these are just because the mock data doesn't have memory. I've also only mocked Mileage here. To me, this is all expected behavior for the mocks.

seesleestak committed:
Refine the sass for custom button pair

seesleestak commented on src/applications/travel-pay/tests/components/complex-claims/pages/IntroductionPage.unit.spec.jsx:
```diff
@@ -97,29 +137,34 @@ describe('Travel Pay – IntroductionPage', () => {
       .exist;
   });
 
-  it('navigates to choose-expense when Start your travel reimbursement claim is clicked', () => {
```
Idk, I may modify this to try and address the first part of [this](https://github.com/department-of-veterans-affairs/vets-website/pull/39673#issuecomment-3504784349)

seesleestak committed:
Add more mocks

seesleestak committed:
Add cancel expense modal

seesleestak committed:
Wire up expense page with expense calls

seesleestak committed:
Use cancel modal on Mileage

seesleestak committed:
Cancel modal tests

seesleestak committed:
Update expense page tests

seesleestak committed:
Merge branch 'main' into travel-pay/complex-claims-redux-setup

seesleestak committed:
Return response from createComplexClaim action

seesleestak committed:
Rework navigation and claim creation on intro page

seesleestak committed:
Unit test updates

seesleestak committed:
[#124445] Travel Pay, Complex claims unsupported mileage page (#39871)

* Add new route

* Add unsupported mileage page

* Add navigate from mileage

* Add test

* Content update for unsupported mileage page

* Test update

* Match route definition in unit test

seesleestak committed:
Fix edit in expense page route definition

seesleestak committed:
Handle getting expense type from edit in expense page

seesleestak committed:
Merge branch 'travel-pay/complex-claims-redux-setup' of github.com:department-of-veterans-affairs/vets-website into travel-pay/complex-claims-redux-setup

pmclaren19 reviewed (dismissed):
- Back button the choose expense page is now fixed and navigates correctly!
- Mileage page back buttons looks correct 
- The add expense page and add mileage expense page/buttons appear to work correctly
- The edit expense page and edit mileage expense page/buttons appear to work correctly
- The delete expense and delete mileage expense buttons/ modal appear to work correctly

Great job @seesleestak

kjduensing reviewed (dismissed):
Nice job.

seesleestak committed:
Use regex match instead of nested find for expense type route matching

pmclaren19 commented on src/applications/travel-pay/components/complex-claims/pages/Mileage.jsx:
```diff
@@ -42,23 +49,70 @@ const Mileage = () => {
+      departureAddress === 'another-address' ||
+      tripType === TRIP_TYPES.ONE_WAY.value
+    ) {
+      navigate(`/file-new-claim/${apptId}/${claimId}/unsupported`);
```
👍
